[CUTLASS] Lift shape & stride information as kernel args

ColinPeppler opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #138611



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @desertfire @chauhang @aakhundov

Differential Revision: [D64773324](https://our.internmc.facebook.com/intern/diff/D64773324)

ColinPeppler committed:
[CUTLASS] Support re-running with different shapes

[ghstack-poisoned]

ColinPeppler commented:
@ColinPeppler has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64773324).

ColinPeppler committed:
Update on "(wip)[CUTLASS] Support re-running with different shapes"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang aakhundov

Differential Revision: [D64773324](https://our.internmc.facebook.com/intern/diff/D64773324)

[ghstack-poisoned]

ColinPeppler committed:
Update on "(wip)[CUTLASS] Support re-running with different shapes"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang aakhundov

Differential Revision: [D64773324](https://our.internmc.facebook.com/intern/diff/D64773324)

[ghstack-poisoned]

ColinPeppler committed:
Update on "[CUTLASS] Lift shape & stride information as kernel args"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang aakhundov

Differential Revision: [D64773324](https://our.internmc.facebook.com/intern/diff/D64773324)

[ghstack-poisoned]

ColinPeppler committed:
Update on "[CUTLASS] Lift shape & stride information as kernel args"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang aakhundov

Differential Revision: [D64773324](https://our.internmc.facebook.com/intern/diff/D64773324)

[ghstack-poisoned]

ColinPeppler committed:
Update on "[CUTLASS] Lift shape & stride information as kernel args"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang aakhundov

Differential Revision: [D64773324](https://our.internmc.facebook.com/intern/diff/D64773324)

[ghstack-poisoned]

ColinPeppler committed:
Update on "[CUTLASS] Lift shape & stride information as kernel args"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang aakhundov

Differential Revision: [D64773324](https://our.internmc.facebook.com/intern/diff/D64773324)

[ghstack-poisoned]

ColinPeppler committed:
Update on "[CUTLASS] Lift shape & stride information as kernel args"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang aakhundov

Differential Revision: [D64773324](https://our.internmc.facebook.com/intern/diff/D64773324)

[ghstack-poisoned]

chenyang78 reviewed (approved):
left a minor comment. LGTM, thanks!

ColinPeppler commented:
@ColinPeppler has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64773324).

ColinPeppler commented on torch/_inductor/codegen/cuda/cutlass_utils.py:
```diff
@@ -297,29 +298,29 @@ def get_max_alignment(inductor_layout: Layout) -> int:
-                for dim in range(len(size))
-            ):
-                return alignment
+    alignments = get_alignments(dtype)
+    for alignment in alignments:
+        if not a_factor_of(size[contiguous_dim], alignment) or not a_factor_of(
+            offset, alignment
+        ):
+            continue
+        if all(
+            (dim == contiguous_dim)
+            or a_factor_of(inductor_layout.stride[dim], alignment)
+            for dim in range(len(size))
+        ):
+            return alignment
```
@tenpercent

ColinPeppler commented on torch/_inductor/codegen/cuda/cutlass_utils.py:
```diff
@@ -297,29 +298,29 @@ def get_max_alignment(inductor_layout: Layout) -> int:
+        rem = sympy.Mod(x, alignment)
+        return V.graph.sizevars.evaluate_expr(sympy.Eq(rem, 0))
+
     try:
         contiguous_dim = inductor_layout.stride.index(1)
     except ValueError:
         # No dim with stride 1 found, return 1
         return 1
```
I think this way of finding contig dim isn't dynamic shape aware.

ColinPeppler replied:
this is okay actually since we must find a 1 in the stride

ColinPeppler commented on torch/_inductor/codegen/cuda/cuda_kernel.py:
```diff
@@ -38,13 +52,100 @@ def _normalize_idx(index: int, total_length: int) -> int:
     return index if index >= 0 else index + total_length
 
 
+ValidSymbols = Literal["M", "N", "K", "lda", "ldb", "ldc", "ldd"]
```
good call, will address in a refactor!

ColinPeppler commented:
@pytorchbot merge

ColinPeppler committed:
Update on "[CUTLASS] Lift shape & stride information as kernel args"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang aakhundov

Differential Revision: [D64773324](https://our.internmc.facebook.com/intern/diff/D64773324)

[ghstack-poisoned]

ColinPeppler committed:
Update on "[CUTLASS] Lift shape & stride information as kernel args"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang aakhundov

Differential Revision: [D64773324](https://our.internmc.facebook.com/intern/diff/D64773324)

[ghstack-poisoned]

ColinPeppler commented:
@ColinPeppler has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64773324).

ColinPeppler commented:
@pytorchbot merge
